### DISCUSSION TITLE
Adding ports to imported gds files

### DISCUSSION THREAD

**Discussion by mcrobbins** [gdsfactory/gdsfactory#1600] [Q&A] [answered]
Hi all,

I want to be able to add ports to imported gds files, either by automatically detecting rectangular edges or by defining rectangles in another port_markers layer at the center of the locations where I want the port defined. 

The "Import PDK" documentation has some information on this in the add_ports from pins section but I'm having issues getting this to run successfully. When running something like this: 

`c3 = gf.import_gds(gdspath, decorator=gf.add_ports.add_ports_from_markers_inside)
c3`

to my imported gds which contains rectangles where I want ports in a layer labelled "port_markers" the code runs successfully but no ports are produced. 

I've used help(add_ports) to try to understand things better, but when I try to run as a decorator with more inputs: 

`ysplit = c_split.add_ref(gf.import_gds("Y_splitter_withports.gds", read_metadata=True, decorator = gf.add_ports.add_ports_from_markers_center(pin_layer=(3,0),port_layer=(1,1),port_type="optical")))`

I receive 'TypeError: add_ports_from_markers_center() missing 1 required positional argument: 'component'' 
If I try to run not as a decorator in a seperate line like this, 

`gf.add_ports.add_ports_from_markers_center(component=ysplit ,pin_layer=(3,0),port_layer=(1,1),port_type="optical")`

I get this: 

```
File "/Users/matthewrobbins/mambaforge/envs/tidy3d/lib/python3.11/site-packages/gdsfactory/port.py", line 404, in read_port_markers
    return component.extract(layers=layers)
           ^^^^^^^^^^^^^^^^^
AttributeError: 'ComponentReference' object has no attribute 'extract'
```

Does anyone have any suggestions here on what I'm doing wrong? Or if there is another way I should be doing this? 

Thanks!

**Comment by joamatab** [comment, marked as answer]
hi @mcrobbins,

can you try?

```python
from functools import partial

add_ports = partial(gf.add_ports.add_ports_from_markers_inside, pin_layer=(3,0)
c3 = gf.import_gds(gdspath, decorator=add_ports) 

```

**Reply by mcrobbins** [reply]
That worked! Thanks a lot!
